### PR TITLE
Fixed issue #1116, where gradients were not tracked after moving a tensor to CUDA.

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
The problem was that if a tensor was first created with 'requires_grad' on CPU, the moved to GPU, the gradient tensor would remain on CPU as far as the C++ object was concerned but remained uninitialized and therefore appear as 'null' in C#. 

The fix is not to set `requires_grad` on the tensor until it has already been moved to GPU.

### CODE REVIEW

File: src/Native/LibTorchSharp/THSTensor.cpp [hunk 1]
```diff
@@ -1771,9 +1771,9 @@ void THSTensor_topk(const Tensor tensor, Tensor* (*allocator)(size_t length), co
 {
     CATCH(
         auto topk = tensor->topk(k, dim, largest, sorted);
-    Tensor * result = allocator(2);
-    result[0] = new torch::Tensor(std::get<0>(topk));
-    result[1] = new torch::Tensor(std::get<1>(topk));
+        Tensor * result = allocator(2);
```
Comment (stephentoub) [inline src/Native/LibTorchSharp/THSTensor.cpp:1774]: Style nit: just looking at pointer use above, the style seems to be to not have a space before the asterisk, e.g.
```suggestion
        Tensor* result = allocator(2);
```

File: src/Native/LibTorchSharp/THSTensor.cpp [hunk 2]
```diff
@@ -1879,10 +1879,10 @@ void THSTensor_unbind(const Tensor tensor, Tensor* (*allocator)(size_t length),
 {
     CATCH(
         auto res = tensor->unbind(dim);
-    const size_t sz = res.size();
-    Tensor * result = allocator(sz);
-    for (size_t i = 0; i < sz; i++)
-        result[i] = new torch::Tensor(res[i]);
+        const size_t sz = res.size();
+        Tensor * result = allocator(sz);
```
Comment (stephentoub) [inline src/Native/LibTorchSharp/THSTensor.cpp:1883]: Nit: same here?
```suggestion
        Tensor* result = allocator(sz);
```

File: src/Native/LibTorchSharp/THSTensorFactories.cpp [hunk 3]
```diff
@@ -170,43 +177,100 @@ Tensor THSTensor_new(
...
+        .requires_grad(requires_grad && !move);
+
+    CATCH(
+        auto res = torch::from_blob(data, at::ArrayRef<int64_t>(sizes, szlength), deleter, options);
+        if (move) // Not CPU
+        {
+            auto device = c10::Device((c10::DeviceType)device_type, (c10::DeviceIndex)device_index);
+            res = convert
+                ? res.to(device, at::ScalarType(dtype), false, false).set_requires_grad(requires_grad)
+                : res.to(device, false, false).set_requires_grad(requires_grad);
```
Comment (stephentoub) [inline src/Native/LibTorchSharp/THSTensorFactories.cpp:200]: Just a style question. We prefer this over e.g.
```suggestion
            res = convert
                ? res.to(device, at::ScalarType(dtype), false, false)
                : res.to(device, false, false);
            res = res.set_requires_grad(requires_grad);
```

**Review by stephentoub — APPROVED** [review]
New to the codebase, but LGTM

File: src/Native/LibTorchSharp/THSTensorFactories.cpp (hunk 3 above)
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSTensorFactories.cpp:200]: Well, to a certain point (line length being reasonable), I personally like it, feels more functional, but I won't pretend that I'm consistent about it.

File: src/Native/LibTorchSharp/THSTensor.cpp (hunk 1 above)
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSTensor.cpp]: K. Will fix.
